crates/core/src/database_instance_context_controller.rs

```diff
@@ -1,63 +1,139 @@
     }
 
+    /// Get the database instance state, or initialize it if it is not present.
+    ///
+    /// If the instance is not initialized yet, this method will block until
+    /// `F` returns. It will, however, release internal locks so calls to other
+    /// methods will not block.
+    ///
+    /// After this method returns, the instance state becomes managed by the
+    /// controller until it is removed by calling [`Self::remove`].
+    ///
+    /// Note that [`Self::remove`] must be called eventually, even if this
+    /// method returns an `Err` result: in this case, [`Self::get`] returns
+    /// `None` (as one would expect), but the given `database_instance_id` is
+    /// nevertheless known to the controller.
     #[tracing::instrument(skip_all)]
-    pub fn insert(&self, database_instance_context: Arc<DatabaseInstanceContext>, scheduler: Scheduler) {
-        let database_instance_id = database_instance_context.database_instance_id;
-        let mut contexts = self.contexts.lock().unwrap();
-        contexts.insert(database_instance_id, (database_instance_context, scheduler));
+    pub fn get_or_try_init<F, E>(
+        &self,
+        database_instance_id: u64,
+        f: F,
+    ) -> Result<(Arc<DatabaseInstanceContext>, Scheduler), E>
+    where
+        F: FnOnce() -> Result<(DatabaseInstanceContext, Scheduler), E>,
+    {
+        let cell = {
+            let mut guard = self.contexts.lock().unwrap();
+            let cell = guard
+                .entry(database_instance_id)
+                .or_insert_with(|| Arc::new(OnceCell::new()));
+            Arc::clone(cell)
+        };
+        cell.get_or_try_init(|| {
+            let (dbic, scheduler) = f()?;
+            Ok((Arc::new(dbic), scheduler))
+        })
+        .map(|(dbic, scheduler)| (dbic.clone(), scheduler.clone()))
     }
 
+    /// Remove and return the state corresponding to `database_instance_id`.
+    ///
+    /// Returns `None` if either the state is not known, or was not properly
+    /// initialized (i.e. [`Self::get_or_try_init`] returned an error).
+    ///
+    /// This method may block if the instance state is currently being
+    /// initialized via [`Self::get_or_try_init`].
     #[tracing::instrument(skip_all)]
     pub fn remove(&self, database_instance_id: u64) -> Option<(Arc<DatabaseInstanceContext>, Scheduler)> {
         let mut contexts = self.contexts.lock().unwrap();
-        contexts.remove(&database_instance_id)
+        let arc = contexts.remove(&database_instance_id)?;
+        match Arc::try_unwrap(arc) {
+            Ok(cell) => cell.into_inner(),
+            Err(arc) => {
+                // If the `Arc`'s refcount is > 1, another thread is currently
+                // executing the `get_or_try_init` closure. Wait until it
+                // completes (instead of returning `None`), as callers rely on
+                // calling `scheduler.clear()`.
+                let (dbic, scheduler) = arc.wait();
```

Comment:
This is a bit finicky: ideally, calling `remove` would give us ownership of the values, because what we want most of the time is to release all resources.

I'm wondering if we'd want to unify the concepts of `HostController` and `DatabaseInstanceContextContoller`, because the former is actually what controls database lifecycle.

Reply:
To be clear: `Arc::try_unwrap` _should_ usually succeed, because current callers remove the instance context from `HostController` before calling `remove()`.